### DISCUSSION TITLE
Testing

### DISCUSSION THREAD

**Discussion by lloricode** [rappasoft/laravel-livewire-tables#271] [General]
Hi @rappasoft @bdelamatre

We have something want to achieve like in testing

how can we test specific result of row or cells?

we been using this
```php
        $this->get('/backend')
            ->assertOk()
            ->assertSeeLivewire($this->livewireIndexTable);
```
while `$livewireIndexTable` is a string class name of DataTableComponent

but seem we can not check if the result is ok

thank you

**Comment by rappasoft** [comment]
You can take a look at what we have testing wise so far for some ideas. We're still working on it ourselves. You can use all the default Livewire testing methods too.

**Comment by bryanbeginly** [comment]
I know this is an older discussion, but I've just started using v3 of the livewire tables, and I have some testing questions as well. I've been trying to test a table with some custom SelectFiters. I've got the test running with the filters pre set, but when I call the render() method in the test, and try to browse the rows, they don't seem to be filtering at all. I'm wondering if there are any particular examples or methods I should work with to execute a successfully test of the table component with filters, sorts, and searches. Happy to share the examples I am working with to help debug if required.

**Reply by lrljoe** [reply]
So, there are extremely comprehensive tests in the package, we cover around 90% of the codebase in terms of tests (which increases with every release).

Typically with packages (for example - Laravel, Livewire, Laravel-Livewire-Tables), adding your own tests, to your own project, isn't considered best practice.
Instead rely on the package tests!

Please feel free to reach out on the official Discord if you want to discuss further, as I think this will probably be a length back and forth that may be easier to do on there!

**Reply by bryanbeginly** [reply]
I see the tests in the code base, and that's great. I'm not trying to recreate the tests, or retest the code from the package. I'm trying to create a unit test to test the table i built with it's custom filters, searches and sorts.  For example, I built a table with some custom search filters to filter my model by location, and by user. I woudl expect that I can define a livewire test similar to the following:

```php 
$component = Livewire::withQueryParams(
                [
                    'table' => [
                        'filters' => [
                            'location' => $locationId,
                        ],
                    ],
                ]
            )
            ->test(MyClass::class)
            ->call('render');
```
And then retrieve the $rows somehow to analyze that they have indeed come back as expected. What i'm attempting to do is get the rows as follows:

```php
$rows = $component->viewData('rows');
```
However, the rows aren't filtered as expected. I've also tried setting the filters array manually and then rendering, but that also doesn't appear to work. The component returns the model data as I expect, but doesn't seem to do any of the filtering functionality.

What I'm trying to work out are what methods can I use to render and retrieve the rows so that I can test my custom logic.

**Reply by lrljoe** [reply]
So $rows isn't passed into the view, it's a computed property, as that's far more efficient.

I'd just use 
```
assertSee
```
and
```
assertDontSee
```
For testing a filter/search

However, should you really want the view to contain $rows, you could call

```useComputedPropertiesDisabled```
This would force the table to run the queries before rendering.

It's really inefficient in comparison tho!
